crates/motsu/src/context.rs

```diff
@@ -90,23 +90,33 @@ impl Context {
         self.storage().contract_address.replace(address)
     }
 
-    /// Set message value to `value` and return the previous value if any.
-    pub(crate) fn set_msg_value(self, value: U256) -> Option<U256> {
-        self.storage().msg_value.replace(value)
+    /// Set an optional message value to `value` and return the previous value
+    /// if any.
+    ///
+    /// Setting `value` to `None` will effectively clear the message value, e.g.
+    /// for non "payable" call.
+    pub(crate) fn set_optional_msg_value(
+        self,
+        value: Option<U256>,
+    ) -> Option<U256> {
+        let msg_value = &mut self.storage().msg_value;
+        let previous = msg_value.take();
+        *msg_value = value;
+        previous
     }
```

Comment:
```suggestion
    pub(crate) fn set_optional_msg_value(self, value: Option<U256>) -> Option<U256> {
        std::mem::replace(&mut self.storage().msg_value, value)
    }
```
This is why I think `fn replace_msg_value` would be more idiomatic